client/gutenberg/editor/utils.js

```diff
@@ -12,20 +13,49 @@ import debugFactory from 'debug';
 
 const debug = debugFactory( 'calypso:gutenberg' );
 
-export const overrideAPIPaths = siteSlug => {
-	// TODO: no API support for now. We'll also need to handle authorization here.
-	apiFetch.use( () => noop );
+export class WithAPIMiddleware extends Component {
+	state = { hasMiddleware: false };
 
-	const rootURL = 'https://public-api.wordpress.com/';
-	apiFetch.use( apiFetch.createRootURLMiddleware( rootURL ) );
+	componentDidMount() {
+		const { siteSlug } = this.props;
 
-	// rewrite default API paths to match WP.com equivalents
-	// Example: /wp/v2/posts -> /wp/v2/sites/{siteSlug}/posts
-	apiFetch.use( ( options, next ) => {
-		const wpcomPath = `/wp/v2/sites/${ siteSlug }/` + options.path.replace( '/wp/v2/', '' );
+		if ( ! isEmpty( siteSlug ) ) {
+			this.applyAPIMiddleware( siteSlug );
+		}
+	}
 
-		debug( 'sending API request to: ', wpcomPath );
+	componentDidUpdate() {
+		const { siteSlug } = this.props;
+		const { hasMiddleware } = this.state;
 
-		return next( { ...options, path: wpcomPath } );
-	} );
-};
+		if ( hasMiddleware || isEmpty( siteSlug ) ) {
+			return;
+		}
+
+		this.applyAPIMiddleware( siteSlug );
+	}
+
+	applyAPIMiddleware = siteSlug => {
+		// TODO: no API support for now. We'll also need to handle authorization here.
+		apiFetch.use( () => noop );
+
+		const rootURL = 'https://public-api.wordpress.com/';
+		apiFetch.use( apiFetch.createRootURLMiddleware( rootURL ) );
+
+		// rewrite default API paths to match WP.com equivalents
+		// Example: /wp/v2/posts -> /wp/v2/sites/{siteSlug}/posts
+		apiFetch.use( ( options, next ) => {
+			const wpcomPath = `/wp/v2/sites/${ siteSlug }/` + options.path.replace( '/wp/v2/', '' );
+
+			debug( 'sending API request to: ', wpcomPath );
+
+			return next( { ...options, path: wpcomPath } );
+		} );
+
+		this.setState( { hasMiddleware: true } );
+	};
```

Comment:
if this is going to merge we will want to make sure we coordinate with https://github.com/Automattic/wp-calypso/pull/26937

also I'd love to have this important part of the middleware extracted into its own module and testable, ideally paired with tests 😄 


```js
const middleware = ( options, next ) => {
	if ( options.path.startsWith( '/wp/v2/sites/' ) ) {

	}

	if ( options.path.startsWith( '/wp/v2/me/' ) ) {

	}

	if ( … ) {

	}
}
```

```js
const returner = options => {
	let output;

	middleware( options, nextOutput => output = nextOutput );

	return output;
}

it( 'removes the site slug', () => {
	[
		[ 'http://base/wp/v2/my-blog.wordpress.com/post', 'https://base/wp/v2/post' ],
		…
	].forEach( ( [ in, out ] ) => expect( returner( in ) ).toEqual( out );
} );
```

better tests of course are due but I'm mainly just scribbling down ideas here. I would hope we can follow this important middleware's tests and see all of the transformations clearly and what we should expect from them

Reply:
@dmsnell refactored this as suggested and added some basic tests that are relevant for now. We can add more tests as we further refine the mapping (we are not sure what exactly we'll need to map yet).